### ISSUE TITLE
EtoMesto.ru

### ISSUE THREAD

**Issue by Arekkusu1998** [lovasoa/dezoomify-rs#222]
Good morning, @lovasoa. I hope everything goes well!
I would have a problem dezoomifying [this map](http://www.etomesto.ru/view.php?map=orenburg_orenburgskoe-voysko-1871-1875&key=1&x=55.101479&y=51.762711): as You can see, it's already georeferenced and this makes everything more convenient for my maps' construction hobby. I tried both the Dezoomify Extension and the Generic Dezoomifier, but I was unsuccessful. 
I apologize if this is an issue that has already been raised.

**Comment by lovasoa** [comment]
This issue in dezoomify references the same site: https://github.com/lovasoa/dezoomify/issues/449

**Comment by lovasoa** [comment]
Quickly looking at the tile URL structure, it looks like you should be able to create a tiles.yaml for it. See https://github.com/lovasoa/dezoomify-rs/wiki/Usage-example-for-the-custom-YAML-dezoomer

**Comment by Arekkusu1998** [comment]
I'm doing something wrong ... I'm really mortified, but I don't khow how to solve this on my own. After a few failed attempts, I made a tiles.yaml that dezoomify-rs accepts.
```
url_template: "http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/{{x}}/{{y}}.jpg"

variables:
  - name: x
    from: 0
    to: 18000 # Image width
    step: 200 # Tile width

  - name: y
    from: 0
    to: 11800 # Image height
    step: 200 # Tile height

headers:
  Referer: "http://www.etomesto.ru/view.php?map=orenburg_orenburgskoe-voysko-1871-1875&key=1&x=55.101479&y=51.762711"

title: "Карта Оренбургского казачьего войска 1871 - 1875 годов"
```
However, it repeats the "_network error: HTTP status client error (404 Not Found)_" and doesn't save but a pitch black image.

**Comment by lovasoa** [comment]
It's not you, it always takes a few attempts to find the right values to put in. You should open your browser's network console and observe the values of x and y in the requests it makes. You have to find the min and max values for x and y. For step, it looks like it's just 1. Neighbouring tiles seem to have consecutive values of x and y.

**Comment by Arekkusu1998** [comment]
By changing step to 1, dezoomify-rs remains stuck in the calculation of the 10 (?) tiles; by changing the other values the 404 Error and the black and/or blank image remain. Speaking about Russian websites, by studying and understanding Issue #120 , I was able to get some maps, but here it's different: I don't have a "previous case" on EtoMesto. 
I give up🏳️.
Please, couldn't You be so kind as to create the `tiles.yaml` file 🙇 (from which I will try to understand how to replicate the procedure for other georeferenced EtoMesto maps)?

**Comment by Arekkusu1998** [comment]
Done. Same result.
![Sample](https://github.com/lovasoa/dezoomify-rs/assets/48092608/43b4c970-9c22-4281-b14f-a3bc17c27ba1)

**Comment by lovasoa** [comment]
Isn't your url template wrong ? 
`http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/{{x}}/{{y}}.jpg`

Look at the tile URLs in your browsers's network panel !

![image](https://github.com/lovasoa/dezoomify-rs/assets/552629/af292e7c-1222-4081-94b8-e3b393b51cd5)

**Comment by Arekkusu1998** [comment]
Okay @lovasoa, good and bad news. 
I solved the problem with #120 maps (some changes in https construction) but -alas!!- not with EtoMesto.

**Comment by lovasoa** [comment]
did you fix your url template ?

**Comment by Arekkusu1998** [comment]
Nope, sorry. I've tried various changes (f.e. `http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg`) but the **404 Error** persists. I only managed to find out that the image is made up of 5460 tiles.

**Comment by Arekkusu1998** [comment]
Nope, sorry. I've tried various changes (f.e.` http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg`) but the **404 Error** persists. I only managed to find out that the image has 5460 tiles and the step is 200.

**Comment by Arekkusu1998** [comment]
Nope, sorry. I've tried various changes (f.e.` http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg`) but the **404 Error** persists. I only managed to find that the image has 5460 tiles and the step is 200.
![Sample](https://github.com/lovasoa/dezoomify-rs/assets/48092608/31b3e656-7030-43fa-b912-583d9e1cea94)

**Comment by Arekkusu1998** [comment]
Nope, sorry. I've tried various changes (f.e.` http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg`) but the **404 Error** persists. I only managed to find that the image has 5460 tiles and the step is 200.
![Sample](https://github.com/lovasoa/dezoomify-rs/assets/48092608/31b3e656-7030-43fa-b912-583d9e1cea94)

**Comment by lovasoa** [comment]
0/5640 ?! Aren't you requesting too many tiles ? Are you sure you found the correct start and end values for x and y ?

**Comment by Arekkusu1998** [comment]
True, it's completely incongruous. Since the size of each tile is 200px, by mathematical calculation I can deduce that x goes from 1 to 90 (18000px), y goes from 1 to 49 (11800px), for a total of 5310 tiles. However, even so, that damned **404 Error** persists🤯.

**Comment by lovasoa** [comment]
Did you check that x goes from 1 to 90 and y goes from 1 to 49 ? You can see that in your browser's network inspector.

**Comment by Arekkusu1998** [comment]
I corrected my previous message.

**Comment by Arekkusu1998** [comment]
I corrected my previous message. 
I also changed the file `tiles.yaml` to make it shorter.
```
url_template: "http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg"
x_template: "x * tile_size"
y_template: "y * tile_size"
variables:
  - { name: x, from: 1, to: 90 } # Number of tiles on the x axis
  - { name: y, from: 1, to: 59 } # Number of tiles on the y axis
  - { name: tile_size, value: 200 }
headers:
  Referer: "http://www.etomesto.ru/view.php?map=orenburg_orenburgskoe-voysko-1871-1875&key=1&x=55.101479&y=51.762711"

title: "result"
```

**Comment by lovasoa** [comment]
Where do these 1 and 90 values come from ? You should check what the actual values for min and max tile number are in your browser.

**Comment by Arekkusu1998** [comment]
They are the minimum & maximum values of x in my browser, sure ... if I interpreted Your screenshot correctly.
![287513638-af292e7c-1222-4081-94b8-e3b393b51cd5](https://github.com/lovasoa/dezoomify-rs/assets/48092608/e9275f0c-1bad-4ab8-adc3-e1939bb1debd)
If I need a georeferenced map from EtoMesto, I can't download it **manually** like this: it's tiring.
![Sample](https://github.com/lovasoa/dezoomify-rs/assets/48092608/5cf8d1ca-46fa-467b-954e-1b71120e6d77)

**Comment by Arekkusu1998** [comment]
They are the minimum & maximum values of x in my browser, sure ... if I interpreted Your screenshot correctly.
![287513638-af292e7c-1222-4081-94b8-e3b393b51cd5](https://github.com/lovasoa/dezoomify-rs/assets/48092608/e9275f0c-1bad-4ab8-adc3-e1939bb1debd)
If I need a georeferenced map from EtoMesto, I can't download it **manually** like this: it's _too_ tiring.
![Sample](https://github.com/lovasoa/dezoomify-rs/assets/48092608/5cf8d1ca-46fa-467b-954e-1b71120e6d77)

**Comment by Arekkusu1998** [comment]
They are the min & max values of x in my browser, sure ... if I interpreted your screenshot correctly.
![287513638-af292e7c-1222-4081-94b8-e3b393b51cd5](https://github.com/lovasoa/dezoomify-rs/assets/48092608/e9275f0c-1bad-4ab8-adc3-e1939bb1debd)

**Comment by Arekkusu1998** [comment]
Good and bad news:
 I solved the problem with #120 maps (minor changes in https construction) but not with **EtoMesto.ru**. Since the size of each tile is 200px, by mathematical calculation I can deduce that x goes from 1 to 90 (18000px), y goes from 1 to 59 (11800px), for a total of 5310 tiles. I also make sorther the file `tiles.yaml`.
```
url_template: "http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/9/{{x}}_{{y}}.jpg"
x_template: "x * tile_size"
y_template: "y * tile_size"
variables:
  - { name: x, from: 1, to: 90 } # Number of tiles on the x axis
  - { name: y, from: 1, to: 59 } # Number of tiles on the y axis
  - { name: tile_size, value: 200 }
headers:
  Referer: "http://www.etomesto.ru/view.php?map=orenburg_orenburgskoe-voysko-1871-1875&key=1&x=55.101479&y=51.762711"
```
However, even so, **network error: HTTP status client error (404 Not Found)** persists🤯. 
![Image](https://github.com/lovasoa/dezoomify-rs/assets/48092608/ad9a007f-1322-4422-a895-0a30f52683cc)
I also tried adding the -H argument, referring to the page's `Cookies`, but that doesn't work (the website may interpret that the download isn't from the legitimate viewer). The format of the tiles itself is anomalous, 'cause it contains three numerical parameters instead of two (x and y). 
Finally, a solution was never given to the Issue https://github.com/lovasoa/dezoomify/issues/449

**Comment by lovasoa** [comment]
I finally found the time to look at it closer.

The problem is the part of the url before the x coordinate. Here is how they generate it:

```js
            ft[1] = "c";
            ft[2] = "c";
            ft[3] = "e";
            ft[4] = "a";
            ft[5] = "e";
            ft[6] = "1";
            ft[7] = "8";
            ft[8] = "c";
            ft[9] = "4";
            ft[10] = "d";
            ft[11] = "6";
            ft[12] = "c";
            ft[13] = "c";
            ft[14] = "a";
            ft[15] = "9";
            ft[16] = "c";
            ft[17] = "7";
            ft[18] = "6";
            ft[19] = "1";
            ft[20] = "9";
            ft[21] = "3";
            ft[22] = "b";
            ft[23] = "3";
            ft[24] = "1";
            ft[25] = "8";
            ft[26] = "4";
            ft[27] = "0";
            ft[28] = "3";
            ft[29] = "6";
            ft[30] = "3";
            ft[31] = "c";
            ft[32] = "6";
            ft[33] = "1";
            ft[34] = "e";
            ft[35] = "1";
            ft[36] = "1";
            ft[37] = "a";
            ft[38] = "a";
            ft[39] = "d";
            ft[40] = "d";
            ft[41] = "3";
            ft[42] = "a";
            ft[43] = "1";
            ft[44] = "f";
            ft[45] = "6";
            ft[46] = "d";
            ft[47] = "6";
            ft[48] = "6";
            ft[49] = "f";
            ft[50] = "c";
            ft[51] = "2";
            ft[52] = "9";
            ft[53] = "d";
            ft[54] = "a";
            ft[55] = "b";
            ft[56] = "9";
            ft[57] = "7";
            ft[58] = "6";
            ft[59] = "0";
            ft[60] = "0";
            ft[61] = "7";
            ft[62] = "4";
            ft[63] = "0";
            ft[64] = "e";
            ft[65] = "f";
            ft[66] = "3";
            ft[67] = "7";
            ft[68] = "a";
            ft[69] = "1";
            ft[70] = "7";
            ft[71] = "e";
            ft[72] = "3";
            ft[73] = "d";
            ft[74] = "a";
            ft[75] = "d";
            ft[76] = "f";
            ft[77] = "2";
            ft[78] = "3";
            ft[79] = "d";
            ft[80] = "f";
            ft[81] = "4";
            ft[82] = "9";
            ft[83] = "f";
            ft[84] = "6";
            ft[85] = "3";
            ft[86] = "9";
            ft[87] = "c";
            ft[88] = "2";
            ft[89] = "7";
            ft[90] = "8";


[...]

                            var text = text + "<img src=" + host + "/orenburg/orenburgskoe-voysko-1871-1875/" + ft[ix] + "/" + ix + "_" + iy + ".jpg width=" + tile + " height=" + tile + " border=0 class=tl>";

```

This is not a known dezoomify-rs format, and wouldn't work with the generic dezoomer either.

**Comment by lovasoa** [comment]
I implemented `str::substring` in dezoomify-rs v2.12 to make it possible to handle such cases in tiles.yaml

```yaml
url_template:
  'http://mapt.etomesto.ru/orenburg/orenburgskoe-voysko-1871-1875/{{
    str::substring(" cceae18c4d6cca9c76193b31840363c61e11aadd3a1f6d66fc29dab97600740ef37a17e3dadf23df49f639c278", x, x+1)
  }}/{{x}}_{{y}}.jpg'
x_template: "x * tile_size"
y_template: "y * tile_size"
variables:
  - { name: "x", from: 4, to: 78 }
  - { name: "y", from: 34, to: 55 }
  - { name: "tile_size", value: 200 }
```


![orenburg](https://github.com/lovasoa/dezoomify-rs/assets/552629/a3844a98-e62a-4d15-8362-32a5fd3ac479)

**Comment by lovasoa** [comment]
I released the new version of dezoomify-rs where `str::substring` is available: https://github.com/lovasoa/dezoomify-rs/releases

The long seemingly random string to use can be found by inspecting the page on etomesto.ru:

In the network panel, I found a tile and clicked the "initiator" tab:
![image](https://github.com/lovasoa/dezoomify-rs/assets/552629/21cfe636-bcad-4449-9405-9ef8ae348f75)

then opened the link:

![image](https://github.com/lovasoa/dezoomify-rs/assets/552629/43ec9c45-f8ab-4da6-8497-b17a0349133d)

and scrolled up a bit

![image](https://github.com/lovasoa/dezoomify-rs/assets/552629/88381273-d793-4595-9083-693207e4c50a)


Happy christmas to you ;)
